Add callback function

Redsandro opened the pull request:
Added callback.

Redsandro committed:
Add callback. Fixes #3

Added callback. Needs testing.

Redsandro committed:
Ignore bad option values

Redsandro committed:
Allow null as options

`null` is an `object`. Cannot `typeof` null keys. Allow passing of `null` to ignore options.

Redsandro committed:
Added test for callback

Redsandro committed:
Bump version

Redsandro commented:
@damirn would you be so kind as to verify if this is to your liking and publish this to `npm`?

Redsandro committed:
Update README.md

Redsandro committed:
Update README.md

Redsandro committed:
Update README.md

damirn commented on index.js:
```diff
@@ -1,16 +1,34 @@
+      }
+    });
+  }
+  
```
an empty line should be removed

Redsandro committed:
remove newline

triviality

Redsandro commented:
Like so?

Feel free to merge and edit it to your styling preferences.

damirn commented:
To be honest, I'm not sure if this approach is good enough. What happens if you feed a huge XML file to rapidx2j and give it a callback? Will the main thread be blocked until parsing is complete or not? Looking at nan's async_pi_estimate example, I'd say a different approach is needed.

Redsandro commented:
I'm afraid you are partially right. It will not block the main thread in the same tick. You can test it out. The CPU will be blocking on next tick, but you can only completely eradicate this by understanding how to have `rapidxml` call back, and I'm not sure how that code works. I think this is the best we can do from the javascript side of things.

`fast-feed` [does something similar](https://github.com/rla/fast-feed/blob/master/index.js#L94), but they do it on the current tick, which is not nice in my opinion.

Irregardless, it's convenient to have a callback. You can directly invoke this in promises this way. If you don't believe deferring the parser to the next tick is good enough for non-blocking purposes, you should see this as a node-style syntactic sugar method, which is still desirable.

We are using it like this, but prefer to be able to use your npm package rather than a hardcoded one. I propose you merge this for the callback convenience, and when you/me/someone has a better idea, we can expand on this functionality. 

This is a step in the right direction. Not a complete solution. That would require a change in the C++ code. But it would be API compatible with this.

Redsandro committed:
Blocking test

You can see that the current execution tick is not blocking.
Next tick is blocking.

That makes this callback a half-solution (better than none in certain cases)

Redsandro commented:
Added a test to show you this

> I'm afraid you are partially right. It will not block the main thread in the same tick. You can test it out. The CPU will be blocking on next tick, but you can only completely eradicate this by understanding how to have rapidxml call back, and I'm not sure how that code works. I think this is the best we can do from the javascript side of things.

damirn commented:
I'll try to complete my work in progress tomorrow and will share it with you asap.

Redsandro commented:
I am curiously looking forward.
